### PR TITLE
refactor/error codes

### PR DESCRIPTION
**Pull request by RuthDiG**

### CODE REVIEW

**Comment by jkoenig134** [comment]
image
you can just change the base branch

**Comment by RuthDiG** [comment]
> image you can just change the base branch

Warum einfach, wenn's auch kompliziert geht..

File: packages/runtime/src/useCases/consumption/notifications/SaveSentNotification.ts [hunk 1]
```diff
@@ -27,14 +27,14 @@ export class SentNotificationUseCase extends UseCase<SentNotificationRequest, Lo
...
     }
 
-    protected async executeInternal(request: SentNotificationRequest): Promise<Result<LocalNotificationDTO, ApplicationError>> {
+    protected async executeInternal(request: SaveSentNotificationRequest): Promise<Result<LocalNotificationDTO, ApplicationError>> {
         const message = await this.messageController.getMessage(CoreId.from(request.messageId));
         if (!message) return Result.fail(RuntimeErrors.general.recordNotFound(Message));
 
         if (!(message.cache!.content instanceof Notification)) return Result.fail(RuntimeErrors.notifications.messageDoesNotContainNotification(message.id));
-        if (!message.isOwn) return Result.fail(RuntimeErrors.notifications.cannotSaveSentNotificationFromPeerMessage());
+        if (!message.isOwn) return Result.fail(RuntimeErrors.notifications.cannotSaveSentNotificationFromPeerMessage(message.id));
```
Comment (Milena-Czierlinski) [inline packages/runtime/src/useCases/consumption/notifications/SaveSentNotification.ts:35]: It would be great if you could add a test for this.

File: packages/consumption/src/modules/notifications/NotificationsController.ts [hunk 2]
```diff
@@ -33,8 +33,8 @@ export class NotificationsController extends ConsumptionBaseController {
         return LocalNotification.from(notification);
     }
 
-    public async sent(message: Message): Promise<LocalNotification> {
-        if (!message.isOwn) throw new Error("Cannot send a Notification from a foreign message.");
+    public async saveSent(message: Message): Promise<LocalNotification> {
```
Comment (britsta) [inline packages/consumption/src/modules/notifications/NotificationsController.ts:36]: Why was the name of the `sent` method of the `NotificationsController` adjusted? Why wasn't the name of the `received` method adjusted as well?

**Review by britsta — COMMENTED** [review]
Version bumps missing.

File: packages/consumption/src/modules/notifications/NotificationsController.ts [hunk 3]
```diff
@@ -33,8 +33,8 @@ export class NotificationsController extends ConsumptionBaseController {
...
+        if (!message.isOwn) throw new Error("Cannot save a LocalNotification from a foreign message.");
```
[the remark this answers is not in the document]
Comment (RuthDiG) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:37]: I would write 'Cannot save a sent LocalNotification from a received Message.', because in this case we create a LocalNotification with LocalNotificationStatus.Sent from an own Message and  in the received method there will be created a LocalNotification with LocalNotificationStatus.Open and receivedByDevice is not undefined. 
So I wouldn't adjust the received method.

File: packages/consumption/src/modules/common/ValidationResult.ts [hunk 4]
```diff
@@ -21,7 +21,13 @@ export abstract class ValidationResult {
 
     public static fromItems(items: ValidationResult[]): ValidationResult {
         return items.some((r) => r.isError())
-            ? ValidationResult.error(new ApplicationError("inheritedFromItem", "Some child items have errors."), items)
+            ? ValidationResult.error(
+                  new ApplicationError(
+                      "error.consumption.validation.inheritedFromItem",
+                      "Some child items have errors. If this error occurred during the creation of a request, call 'canCreate' to get more information."
```
[the remark this answers is not in the document]
Comment (RuthDiG) [reply packages/consumption/src/modules/common/ValidationResult.ts:27]: I'm fine with the first two points but the last one would be too much because when you call accept or reject and canAccept or canReject won't be successful there will be thrown: "Cannot accept the Request with the given parameters. Call 'canAccept' to get more information." or "Cannot reject the Request with the given parameters. Call 'canReject' to get more information."

File: packages/consumption/src/modules/notifications/NotificationsController.ts (hunk 2 above)
Comment (RuthDiG) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:36]: see above.

Comment (jkoenig134) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:36]: I don't see anything above. I don't like the new naming.

**Comment by jkoenig134** [comment]
Who decided on putting all these `s and 's in the error messages? I'm not a fan of this. These symbols are indicating strings, please use them only to surround strings like ids or enums.

File: packages/consumption/src/modules/notifications/NotificationsController.ts (hunk 2 above)
Comment (jkoenig134) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:36]: Additionally: this is nothing which should pop up in a PR that is dedicated to changing error codes and messages.

**Comment by britsta** [comment]
> Who decided on putting all these `s and 's in the error messages? I'm not a fan of this. These symbols are indicating strings, please use them only to surround strings like ids or enums.

Good question! I agree with you.

**Comment by RuthDiG** [comment]
> Who decided on putting all these `s and 's in the error messages? I'm not a fan of this. These symbols are indicating strings, please use them only to surround strings like ids or enums.
I misunderstood what @britsta said here: 
_Also, let's agree on a uniform marking of properties (and their values) of data object types. For example, enclose properties in error messages with backticks `...`, as it is already done in many places._
Because she told me I should do it like in the 'Data Model Overview' and there it was partly wrong.

File: packages/consumption/src/modules/notifications/NotificationsController.ts (hunk 2 above)
Comment (RuthDiG) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:36]: In the other deleted PR @Milena-Czierlinski wrote this:

_The error message seems misleading to me. If I understand it correctly, the use case SentNotification is meant to save a LocalNotification for a Notification that was sent inside a Message by oneself. If the owner of the Message is not the Identity itself, this error is thrown. In this case, I'd expect an error message like The Message ${messageId} was received from a peer, but an own Message is expected here to save its Notification content..

Besides, also the naming of the SentNotificationUseCase seems not ideal to me, as it doesn't appropriately describe what happens inside of it. However, this is probably outside the scope of this PR._

 and after that we agreed about this way and so I changed it.

**Comment by britsta** [comment]
Ohhh I think that was my fault, misunderstanding of Julian's comment!

**Comment by jkoenig134** [comment]
O don't think we should do that here in the errors. It's another thing in the docs.

**Comment by britsta** [comment]
> I don't think we should do that here in the errors. It's another thing in the docs.

Since examples are still the best way to clarify things, which of the variants would you @jkoenig134 prefer?

1. The \`deletionInfo.deletionStatus\` of the LocalAttribute is "DeletionRequestSent".
2. The deletionInfo.deletionStatus of the LocalAttribute is "DeletionRequestSent".
3. The \`deletionInfo.deletionStatus\` of the LocalAttribute is 'DeletionRequestSent'.
4. The deletionInfo.deletionStatus of the LocalAttribute is 'DeletionRequestSent'.

**Comment by jkoenig134** [comment]
IMO 4 is the best for coding.

File: packages/consumption/src/modules/notifications/NotificationsController.ts (hunk 2 above)
Comment (jkoenig134) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:36]: `sent` is within our naming schema (we do the same for requests for example). I stay with my opinion that **if** there would be need for changing this 

1. this is the wrong PR to do so
2. we have to discuss this first

therefore - please undo that everywhere.

File: packages/consumption/test/modules/requests/DecideRequestParametersValidator.test.ts [hunk 5]
```diff
@@ -315,8 +315,10 @@ describe("DecideRequestParametersValidator", function () {
             return;
         }
 
-        expect(validationResult.error.code).toBe("inheritedFromItem");
-        expect(validationResult.error.message).toBe("Some child items have errors.");
+        expect(validationResult.error.code).toBe("error.consumption.validation.inheritedFromItem");
+        expect(validationResult.error.message).toBe(
+            "Some child items have errors. If this error occurred during the specification of a Request, call 'canCreate' to get more information."
+        );
```
Comment (jkoenig134) [inline packages/consumption/test/modules/requests/DecideRequestParametersValidator.test.ts:321]: we have a helper method `expect(result).errorValidationResult(...)`, please use that.

File: packages/consumption/test/modules/requests/IncomingRequestsController.test.ts [hunk 6]
```diff
@@ -364,16 +364,18 @@ describe("IncomingRequestsController", function () {
...
 
             expect(validationResult.items[0].isError()).toBe(false);
 
             expect(validationResult.items[1].isError()).toBe(true);
-            expect((validationResult.items[1] as ErrorValidationResult).error.code).toBe("inheritedFromItem");
-            expect((validationResult.items[1] as ErrorValidationResult).error.message).toBe("Some child items have errors.");
+            expect((validationResult.items[1] as ErrorValidationResult).error.code).toBe("error.consumption.validation.inheritedFromItem");
+            expect((validationResult.items[1] as ErrorValidationResult).error.message).toBe(
+                "Some child items have errors. If this error occurred during the specification of a Request, call 'canCreate' to get more information."
+            );
```
Comment (jkoenig134) [inline packages/consumption/test/modules/requests/IncomingRequestsController.test.ts:378]: see above, same with other occurrences in the file

File: packages/consumption/src/modules/common/ValidationResult.ts [hunk 7]
```diff
@@ -21,7 +21,13 @@ export abstract class ValidationResult {
 
     public static fromItems(items: ValidationResult[]): ValidationResult {
         return items.some((r) => r.isError())
-            ? ValidationResult.error(new ApplicationError("inheritedFromItem", "Some child items have errors."), items)
+            ? ValidationResult.error(
+                  new ApplicationError(
```
[the remark this answers is not in the document]
Comment (RuthDiG) [reply packages/consumption/src/modules/common/ValidationResult.ts:25]: In the case of Accept or Reject I'm fine with that but when you create an outgoing Request you will only get the message of the code 'inheritedFromItem' and no other message...

[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/consumption/src/modules/common/ValidationResult.ts:25]: Ok, then I'm ok with the new version.

File: packages/consumption/src/modules/notifications/NotificationsController.ts (hunk 2 above)
Comment (RuthDiG) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:36]: Ok, than I will undo this.

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 8]
```diff
@@ -15,54 +15,54 @@ class Attributes {
     public successionMustNotChangeKey() {
         return new CoreError(
             "error.consumption.attributes.successionMustNotChangeKey",
-            "The predecessor attribute's key does not match that of the successor. The succession of a relationship attribute must not change the key."
+            "The predecessor Attribute's key does not match that of the successor. The succession of a RelationshipAttribute must not change the key."
```
Comment (britsta) [inline packages/consumption/src/consumption/CoreErrors.ts:18]: "RelationshipAttribute" instead of "Attribute".

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 9]
```diff
@@ -15,54 +15,54 @@ class Attributes {
...
-            "Predecessor source attribute is not succeeded by successor source attribute."
+            "Predecessor sourceAttribute is not succeeded by successor sourceAttribute."
         );
     }
 
     public predecessorSourceContentIsNotEqualToCopyContent() {
         return new CoreError(
             "error.consumption.attributes.predecessorSourceContentIsNotEqualToCopyContent",
-            "Predecessor source attribute content doesn't match predecessor shared attribute copy."
+            "Predecessor sourceAttribute content doesn't match predecessor shared Attribute copy."
```
Comment (britsta) [inline packages/consumption/src/consumption/CoreErrors.ts:44]: "Predecessor sourceAttribute's content does not match **the content of the** predecessor shared Attribute copy."

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 10]
```diff
@@ -15,54 +15,54 @@ class Attributes {
...
         );
     }
 
     public successorSourceContentIsNotEqualToCopyContent() {
         return new CoreError(
             "error.consumption.attributes.successorSourceContentIsNotEqualToCopyContent",
-            "Successor source attribute content doesn't match successor shared attribute copy."
+            "Successor sourceAttribute content doesn't match successor shared Attribute copy."
```
Comment (britsta) [inline packages/consumption/src/consumption/CoreErrors.ts:51]: "Successor sourceAttribute's content does not match **the content of the** successor shared Attribute copy."

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 11]
```diff
@@ -15,54 +15,54 @@ class Attributes {
...
         );
     }
 
     public cannotSucceedChildOfComplexAttribute(parentId: string | CoreId) {
         return new CoreError(
             "error.consumption.attributes.cannotSucceedChildOfComplexAttribute",
-            `The attribute you want to succeed is child attribute of a complex attribute (id: ${parentId}), and cannot be succeeded on its own. Instead, succeed the parent which will implicitly succeed all its children.`
+            `The Attribute you want to succeed is child Attribute of a complex Attribute (id: '${parentId}'), and cannot be succeeded on its own. Instead, succeed the parent which will implicitly succeed all its children.`
```
Comment (britsta) [inline packages/consumption/src/consumption/CoreErrors.ts:58]: "The Attribute you want to succeed is **a** child Attribute of a..."

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 12]
```diff
@@ -224,28 +224,28 @@ class Attributes {
...
-            `The attribute (id: ${attributeId}) is not a third party owned RelationshipAttribute.`
+            `The Attribute (id: '${attributeId}') is not a third party owned RelationshipAttribute.`
         );
     }
 
     public senderIsNotPeerOfSharedAttribute(senderId: string | CoreAddress, attributeId: string | CoreId) {
         return new CoreError(
             "error.consumption.attributes.senderIsNotPeerOfSharedAttribute",
-            `The sender (id: ${senderId}) is not the peer you shared the attribute (id: ${attributeId}) with.`
+            `The sender (id: '${senderId}') is not the peer you shared the attribute (id: '${attributeId}') with.`
```
Comment (britsta) [inline packages/consumption/src/consumption/CoreErrors.ts:248]: - Is this the Sender of a Notification? If so, make this clear.
- "Attribute" instead of "attribute".

File: packages/consumption/src/modules/notifications/NotificationsController.ts [hunk 13]
```diff
@@ -34,7 +34,7 @@ export class NotificationsController extends ConsumptionBaseController {
     }
 
     public async sent(message: Message): Promise<LocalNotification> {
-        if (!message.isOwn) throw new Error("Cannot send a Notification from a foreign message.");
+        if (!message.isOwn) throw new Error("Cannot save a sent LocalNotification from a received Message.");
```
Comment (britsta) [inline packages/consumption/src/modules/notifications/NotificationsController.ts:37]: This is a little bit confusing, maybe rephrase: "Cannot save a LocalNotification with status 'Sent' from a received Message."

File: packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts [hunk 14]
```diff
@@ -66,7 +66,7 @@ describe("End2End Notification via Messages", function () {
         TestNotificationItemProcessor.reset();
     });
 
-    test("sender: sent Notification", async function () {
+    test("sender: save sent Notification", async function () {
```
Comment (britsta) [inline packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts:69]: This is a little bit confusing, maybe rephrase: "sender: save Notification with status 'Sent'"

File: packages/runtime/src/useCases/common/RuntimeErrors.ts [hunk 15]
```diff
@@ -96,18 +96,18 @@ class Relationships {
     public wrongRelationshipStatus(relationshipId: string, status: string): ApplicationError {
         return new ApplicationError(
             "error.runtime.relationships.wrongRelationshipStatus",
-            `The relationship '${relationshipId}' has the wrong status (${status}) to run this operation`
+            `The relationship '${relationshipId}' has the wrong status ('${status}') to run this operation`
```
Comment (britsta) [inline packages/runtime/src/useCases/common/RuntimeErrors.ts:99]: -"Relationship" instead of "relationship".
- Please end this sentence with a full stop.

File: packages/runtime/src/useCases/common/RuntimeErrors.ts [hunk 16]
```diff
@@ -164,29 +167,29 @@ class Attributes {
...
+            `RepositoryAttribute '${repositoryAttributeId.toString()}' has already been shared with peer '${peer.toString()}'. ID of own shared IdentityAttribute: '${ownSharedIdentityAttributeId.toString()}'.`
         );
     }
 
     public noPreviousVersionOfRepositoryAttributeHasBeenSharedWithPeerBefore(repositoryAttributeId: CoreId | string, peer: CoreAddress | string): ApplicationError {
         return new ApplicationError(
             "error.runtime.attributes.noPreviousVersionOfRepositoryAttributeHasBeenSharedWithPeerBefore",
-            `No previous version of repository attribute '${repositoryAttributeId.toString()}' has been shared with peer '${peer.toString()}' before. If you wish to execute an initial sharing of this attribute, use 'ShareRepositoryAttribute'.`
+            `No previous version of RepositoryAttribute '${repositoryAttributeId.toString()}' has been shared with peer '${peer.toString()}' before. If you wish to execute an initial sharing of this Attribute, use ShareRepositoryAttribute.`
```
Comment (britsta) [inline packages/runtime/src/useCases/common/RuntimeErrors.ts:177]: ShareRepositoryAttribute refers to the ShareRepositoryAttributeUseCase? If so, adjust this here appropriately.

File: packages/transport/src/core/CoreErrors.ts [hunk 17]
```diff
@@ -138,24 +134,20 @@ class Files {
...
-    public fileContentUndefined() {
-        return new CoreError("error.transport.files.fileContentUndefined", "The given file content is undefined.");
+        return new CoreError("error.transport.files.invalidMetadata", `The metadata of the File with id '${id}' is invalid.`);
     }
 
     public maxFileSizeExceeded(fileSize: number, platformMaxFileSize: number) {
         return new CoreError(
             "error.transport.files.maxFileSizeExceeded",
-            `The given file content size (${fileSize}) exceeds the max file size the backbone accepts (${platformMaxFileSize}).`
+            `The given file content size (${fileSize}) exceeds the max file size the Backbone accepts (${platformMaxFileSize}).`
```
Comment (britsta) [inline packages/transport/src/core/CoreErrors.ts:143]: "File" instead of "file".

File: packages/runtime/test/consumption/notifications.test.ts [hunk 18]
```diff
@@ -122,12 +135,12 @@ describe("Notifications", () => {
             expect(result).toBeAnError("'LocalNotification' not found.", "error.transport.recordNotFound");
         });
 
-        test("wrong message id for sentNotification", async () => {
+        test("wrong message id for saveSentNotification", async () => {
```
Comment (britsta) [inline packages/runtime/test/consumption/notifications.test.ts:138]: - "Message ID" instead of "message id".
- Rephrase: "saveSentNotification" to "save Notification with status 'Sent'"

File: packages/runtime/test/consumption/notifications.test.ts [hunk 19]
```diff
@@ -122,12 +135,12 @@ describe("Notifications", () => {
...
             const result = await sConsumptionServices.notifications.sentNotification({ messageId: "wrong-id" });
             expect(result).toBeAnError("messageId must match pattern MSG\\[A-Za-z0-9\\]\\{17\\}", "error.runtime.validation.invalidPropertyValue");
         });
 
-        test("not existing message id for sentNotification", async () => {
+        test("not existing message id for saveSentNotification", async () => {
```
Comment (britsta) [inline packages/runtime/test/consumption/notifications.test.ts:143]: - "Message ID" instead of "message id".
- Rephrase: "saveSentNotification" to "save Notification with status 'Sent'"

File: packages/runtime/test/consumption/notifications.test.ts [hunk 20]
```diff
@@ -53,6 +53,19 @@ describe("Notifications", () => {
         expect(getNotificationResult.value.status).toStrictEqual(LocalNotificationStatus.Sent);
     });
 
+    test("sentNotification with error", async () => {
```
Comment (britsta) [inline packages/runtime/test/consumption/notifications.test.ts:56]: "sent Notification" instead of "sentNotification"

File: packages/consumption/src/modules/notifications/NotificationsController.ts [hunk 21]
```diff
@@ -34,7 +34,7 @@ export class NotificationsController extends ConsumptionBaseController {
     }
 
     public async sent(message: Message): Promise<LocalNotification> {
-        if (!message.isOwn) throw new Error("Cannot send a Notification from a foreign message.");
+        if (!message.isOwn) throw new Error("Cannot save a LocalNotification with status 'Sent' from a received Message.");
```
Comment (jkoenig134) [inline packages/consumption/src/modules/notifications/NotificationsController.ts:37]: I think you forgot to undo stuff here

File: packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts [hunk 22]
```diff
@@ -66,7 +66,7 @@ describe("End2End Notification via Messages", function () {
         TestNotificationItemProcessor.reset();
     });
 
-    test("sender: sent Notification", async function () {
+    test("sender: save Notification with status 'Sent'", async function () {
```
Comment (jkoenig134) [inline packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts:69]: I think you forgot to undo stuff here

File: packages/runtime/test/consumption/notifications.test.ts [hunk 23]
```diff
@@ -122,12 +135,12 @@ describe("Notifications", () => {
             expect(result).toBeAnError("'LocalNotification' not found.", "error.transport.recordNotFound");
         });
 
-        test("wrong message id for sentNotification", async () => {
+        test("wrong Message ID for sentNotification with status 'Sent'", async () => {
```
Comment (jkoenig134) [inline packages/runtime/test/consumption/notifications.test.ts:138]: I don't see a status being important in this test?

File: packages/runtime/test/consumption/notifications.test.ts [hunk 24]
```diff
@@ -122,12 +135,12 @@ describe("Notifications", () => {
...
             const result = await sConsumptionServices.notifications.sentNotification({ messageId: "wrong-id" });
             expect(result).toBeAnError("messageId must match pattern MSG\\[A-Za-z0-9\\]\\{17\\}", "error.runtime.validation.invalidPropertyValue");
         });
 
-        test("not existing message id for sentNotification", async () => {
+        test("not existing Message ID for sentNotification with status 'Sent'", async () => {
```
Comment (jkoenig134) [inline packages/runtime/test/consumption/notifications.test.ts:143]: see above.

File: packages/consumption/src/modules/notifications/NotificationsController.ts (hunk 21 above)
Comment (RuthDiG) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:37]: I only changed the Message because the Message before wasn't right. See the discussions before. Or what do you mean?

Comment (jkoenig134) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:37]: Ok, maybe I am just confused. What do you think about: `Cannot mark a LocalNotification as sent from a received Message.`

Comment (RuthDiG) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:37]: Okay so it would be an error message analogous to the one in the handleMessageReceivedEvent.

Comment (jkoenig134) [reply packages/consumption/src/modules/notifications/NotificationsController.ts:37]: yep!

File: packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts (hunk 22 above)
Comment (RuthDiG) [reply packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts:69]: So I would change this to "sender:  mark LocalNotification as sent" in order to be consistent?

Comment (jkoenig134) [reply packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts]: no, `sender: sent Notification` was correct. this test tests the function sentNotification which sets the test name :)

**Review by jkoenig134 — DISMISSED** [review]

File: packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts (hunk 22 above)
Comment (jkoenig134) [reply packages/consumption/test/modules/notifications/NotificationEnd2End.test.ts]: but I'm also ok with your new test name

**Review by jkoenig134 — APPROVED** [review]
